beacon-chain/sync/validate_sync_committee_message.go

```diff
@@ -0,0 +1,310 @@
+	return func(ctx context.Context) pubsub.ValidationResult {
+		_, span := trace.StartSpan(ctx, "sync.ifInvalidSyncMsgTime")
+		defer span.End()
+		// The message's `slot` is for the current slot (with a MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance).
+		if err := altair.ValidateSyncMessageTime(
+			m.Slot,
+			s.cfg.Chain.GenesisTime(),
+			params.BeaconNetworkConfig().MaximumGossipClockDisparity,
+		); err != nil {
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationIgnore
+		}
+		return pubsub.ValidationAccept
+	}
+}
+
+// The `subnet_id` is valid for the given validator. This implies the validator is part of the broader
+// current sync committee along with the correct subcommittee.
+// We are trying to validate that whatever committee indices that were retrieved from our state for this
+// particular validator are indeed valid for this particular topic. Ex: the topic name can be
+// /eth2/b5303f2a/sync_committee_2/ssz_snappy
+// This would mean that only messages meant for subnet 2 are valid. If a validator creates this sync
+// message and broadcasts it into subnet 2, we need to make sure that whatever committee index and
+// resultant subnet that the validator has is valid for this particular topic.
+func (s *Service) rejectIncorrectCommittee(
+	committeeIndices []types.CommitteeIndex, topic string,
+) validationFn {
+	return func(ctx context.Context) pubsub.ValidationResult {
+		_, span := trace.StartSpan(ctx, "sync.ifIncorrectCommittee")
+		defer span.End()
+		isValid := false
+		digest, err := s.forkDigest()
+		if err != nil {
+			traceutil.AnnotateError(span, err)
+			return pubsub.ValidationIgnore
+		}
+
+		format := p2p.GossipTypeMapping[reflect.TypeOf(&ethpb.SyncCommitteeMessage{})]
+		// Validate that the validator is in the correct committee.
+		subCommitteeSize := params.BeaconConfig().SyncCommitteeSize / params.BeaconConfig().SyncCommitteeSubnetCount
+		for _, idx := range committeeIndices {
+			subnet := uint64(idx) / subCommitteeSize
+			if strings.HasPrefix(topic, fmt.Sprintf(format, digest, subnet)) {
+				isValid = true
+				break
+			}
+		}
+		if !isValid {
+			return pubsub.ValidationReject
+		}
+		return pubsub.ValidationAccept
+	}
+}
+
+// There has been no other valid sync committee signature for the declared `slot`, `validator_index`,
+// and `subcommittee_index`. In the event of `validator_index` belongs to multiple subnets, as long
+// as one subnet has not been seen, we should let it in.
+func (s *Service) ignoreHasSeenSyncMsg(
+	m *ethpb.SyncCommitteeMessage, committeeIndices []types.CommitteeIndex,
+) validationFn {
+	return func(ctx context.Context) pubsub.ValidationResult {
+		var isValid bool
+		subCommitteeSize := params.BeaconConfig().SyncCommitteeSize / params.BeaconConfig().SyncCommitteeSubnetCount
+		for _, idx := range committeeIndices {
+			subnet := uint64(idx) / subCommitteeSize
+			if s.hasSeenSyncMessageIndexSlot(m.Slot, m.ValidatorIndex, subnet) {
```

Comment:
```suggestion
			if !s.hasSeenSyncMessageIndexSlot(m.Slot, m.ValidatorIndex, subnet) {
			isValid := true
			break
```
you can simplify the conditional by breaking early if we haven't observed it in the cache.